cpp/src/arrow/engine/substrait/relation_internal.cc

```diff
@@ -49,6 +50,52 @@ Status CheckRelCommon(const RelMessage& rel) {
   return Status::OK();
 }
 
+Result<fs::FileInfoVector> GetGlobFiles(const std::shared_ptr<fs::FileSystem>& filesystem,
+                                        const std::string& glob) {
+  fs::FileInfoVector results, temp;
+  fs::FileSelector selector;
+  std::string cur;
+
+  auto split_path = fs::internal::SplitAbstractPath(glob, '/');
+  ARROW_ASSIGN_OR_RAISE(auto file, filesystem->GetFileInfo("/"));
+  results.push_back(std::move(file));
+
+  for (size_t i = 0; i < split_path.size(); i++) {
+    if (split_path[i].find_first_of("*?") == std::string::npos) {
+      if (cur.empty())
+        cur = split_path[i];
+      else
+        cur = fs::internal::ConcatAbstractPath(cur, split_path[i]);
+      continue;
+    } else {
+      for (auto res : results) {
+        if (res.type() != fs::FileType::Directory) continue;
+        selector.base_dir = res.path() + cur;
+        ARROW_ASSIGN_OR_RAISE(auto entries, filesystem->GetFileInfo(selector));
```

Comment:
Note that instead of issuing a blocking call for each selector, it would probably be more efficient to use the async filesystem APIs. This can be left for another JIRA, though.

Reply:
https://issues.apache.org/jira/browse/ARROW-16459